go-client/client/fieldmanager.go

```diff
@@ -0,0 +1,368 @@
+func (fs *fieldStream) FetchRepeating(stream chan<- respFMExec) {
+	go func() {
+		for {
+			changes, err := fs.fieldsClient.Recv()
+			stream <- respFMExec{changes: changes, err: err}
+
+			if err != nil {
+				// The error will be handled by the fieldManagerExecutor.
+				fs.Close()
+				return
+			}
+		}
+	}()
+}
+
+// Close closes the field stream and stops any existing FetchRepeating requests.
+func (fs *fieldStream) Close() {
+	if fs.cancel != nil {
+		fs.cancel()
+		fs.cancel = nil
+	}
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+// isCanceledError returns true if the error is a gRPC Canceled error.
+func isCanceledError(err error) bool {
+	status, ok := status.FromError(err)
+	return ok && status.Code() == codes.Canceled
+}
+
+// fieldManagerExecutor handles requests from a fieldManager and performs the appropriate actions.
+// It controls the state of FetchTablesRepeating calls and holds the list of tables that can be opened.
+type fieldManagerExecutor struct {
+	client *Client
+	tables map[fieldId]*TableHandle // A map of tables that can be opened using OpenTable
+	fs     *fieldStream             // The current stream for a FetchRepeating request. This is nil if no FetchRepeating request is running.
+
+	chanClose               <-chan reqClose
+	chanOpenTables          <-chan reqListOpenableTables
+	chanGetTable            <-chan reqGetTable
+	chanFMExec              <-chan reqFMExec
+	chanFetchRepeating      <-chan reqFetchRepeating
+	chanFetchRepeatingError chan<- error // Errors coming from FetchRepeating requests will be sent to this channel.
+}
+
+// loop starts up an executor loop, which will handle requests
+// from the executor's channels and perform the appropriate actions.
+// It will run forever until chanClose is closed.
+func (fme *fieldManagerExecutor) loop() {
+	// This channel is used to send/receive responses from FetchRepeating requests.
+	chanChanges := make(chan respFMExec)
+
+	defer func() {
+		if fme.fs != nil {
+			fme.fs.Close()
+			fme.fs = nil
+		}
+
+		if fme.chanFetchRepeatingError != nil {
+			close(fme.chanFetchRepeatingError)
+			fme.chanFetchRepeatingError = nil
+		}
+	}()
+
+	for {
+		select {
+		case <-fme.chanClose:
+			return
+		case req := <-fme.chanOpenTables:
+			req.out <- fme.listOpenableTables()
+		case req := <-fme.chanGetTable:
+			tbl, ok := fme.tables[req.id]
+			req.out <- respGetTable{table: tbl, ok: ok}
+		case req := <-fme.chanFMExec:
+			changes, err := req.f()
+			msg := respFMExec{changes: changes, err: err}
+			// The ExecAndUpdate call will handle the error for us.
+			req.out <- msg
+			if err == nil {
+				fme.handleFieldChanges(changes)
+			}
+		case req := <-fme.chanFetchRepeating:
+			if fme.fs != nil {
+				fme.fs.Close()
+				fme.fs = nil
+			}
+
+			if fme.chanFetchRepeatingError != nil {
+				close(fme.chanFetchRepeatingError)
+				fme.chanFetchRepeatingError = nil
+			}
```

Comment:
This code looks duped from above.  DRY.  Just make one func and use it in both places.